cocos/core/geometry/spline.ts

```diff
@@ -0,0 +1,292 @@
+/*
+ Copyright (c) 2022 Xiamen Yaji Software Co., Ltd.
+
+ https://www.cocos.com/
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated engine source code (the "Software"), a limited,
+ worldwide, royalty-free, non-assignable, revocable and non-exclusive license
+ to use Cocos Creator solely to develop games on your target platforms. You shall
+ not use Cocos Creator software for developing other software or tools that's
+ used for developing games. You are not granted to publish, distribute,
+ sublicense, and/or sell copies of Cocos Creator.
+
+ The software or tools in this License Agreement are licensed, not sold.
+ Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ */
+
+/**
+ * @packageDocumentation
+ * @module geometry
+ */
+
+import { assertIsTrue } from '../data/utils/asserts';
+import { clamp, Vec3 } from '../math';
+import { warnID } from '../platform/debug';
+import enums from './enums';
+
+export enum SplineMode {
+    /**
+     * Broken line:
+     * Each knot is connected with a straight line from the beginning to the end to form a curve. At least two knots.
+     */
+    LINEAR = 0,
+
+    /**
+     * Piecewise Bezier curve:
+     * Every four knots form a curve. Total knots number must be a multiple of 4.
+     * Each curve passes only the first and fourth knots, and does not pass through the middle two control knots.
+     * 
+     * If you need a whole continuous curve:
+     * (1) Suppose the four knots of the previous curve are A, B, C, D
+     * (2) The four knots of the next curve must be D, E, F, G
+     * (3) C and E need to be symmetrical about D
+     */
+    BEZIER = 1,
+
+    /**
+     * Catmull Rom curve:
+     * All knots(including start & end knots) form a whole continuous curve. At least two knots.
+     * The whole curve passes through all knots.
+     */
+    CATMULL_ROM = 2
+};
+
+const SPLINE_WHOLE_INDEX = 0xffffffff;
+
+const _v0 = new Vec3();
+const _v1 = new Vec3();
+const _v2 = new Vec3();
+const _v3 = new Vec3();
+
+/**
+ * @en
+ * Basic Geometry: Spline.
+ * @zh
+ * 基础几何 Spline
+ */
+
+export class Spline {
+    private readonly _type: number;
+    private _mode: SplineMode = SplineMode.CATMULL_ROM;
+    private _knots: Vec3[] = [];
+
+    public constructor(mode: SplineMode = SplineMode.CATMULL_ROM, knots: Vec3[] = []) {
```

Comment:
As there is a create method, should make the constructor public?

Reply:
This is consistent with other geometry.

Reply:
It is a new class, can use correct way to handle it. And it doesn't break anything.